File: services/catalog/internal/apps/catalog/controllers/product.go
package catalogControllers

import (
	"github.com/gofiber/fiber/v2"
	"github.com/usegro/services/catalog/internal/apps/catalog/dto"
	catalogServices "github.com/usegro/services/catalog/internal/apps/catalog/services"
	"github.com/usegro/services/catalog/internal/interface/response"
)

type ProductController struct {
	service *catalogServices.ProductService
}

func NewProductController(service *catalogServices.ProductService) *ProductController {
	return &ProductController{service: service}
}

func (pc *ProductController) CreateProduct(c *fiber.Ctx) error {
	crmID := c.Locals("crmID").(string)

	var body dto.CreateProductDTO
	if err := c.BodyParser(&body); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(response.CommonResponse{
			ResponseCode:    response.BAD_REQUEST,
			ResponseMessage: "Invalid request body",
		})
	}

	if body.Name == "" {
		return c.Status(fiber.StatusBadRequest).JSON(response.CommonResponse{
			ResponseCode:    response.BAD_REQUEST,
			ResponseMessage: "Product name is required",
		})
	}

	product, err := pc.service.CreateProduct(c.Context(), crmID, body)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(response.CommonResponse{
			ResponseCode:    response.BAD_REQUEST,
			ResponseMessage: err.Error(),
		})
	}

	return c.Status(fiber.StatusCreated).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_CREATED,
		ResponseMessage: "Product created successfully",
		Data:            product,
	})
}

func (pc *ProductController) ListProducts(c *fiber.Ctx) error {
	crmID := c.Locals("crmID").(string)
	search := c.Query("search")
	status := c.Query("status")
	page := c.QueryInt("page", 1)
	limit := c.QueryInt("limit", 10)

	products, err := pc.service.ListProducts(c.Context(), crmID, search, status, page, limit)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(response.CommonResponse{
			ResponseCode:    response.BAD_REQUEST,
			ResponseMessage: err.Error(),
		})
	}

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_FETCHED,
		ResponseMessage: "Products retrieved successfully",
		Data:            products,
	})
}

func (pc *ProductController) GetProduct(c *fiber.Ctx) error {
	crmID := c.Locals("crmID").(string)
	productID := c.Params("id")

	if productID == "" {
		return c.Status(fiber.StatusBadRequest).JSON(response.CommonResponse{
			ResponseCode:    response.BAD_REQUEST,
			ResponseMessage: "Product ID is required",
		})
	}

	product, err := pc.service.GetProduct(c.Context(), crmID, productID)
	if err != nil {
		return c.Status(fiber.StatusNotFound).JSON(response.CommonResponse{
			ResponseCode:    response.RESOURCE_NOT_FOUND,
			ResponseMessage: err.Error(),
		})
	}

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_FETCHED,
		ResponseMessage: "Product retrieved successfully",
		Data:            product,
	})
}

func (pc *ProductController) UpdateProduct(c *fiber.Ctx) error {
	crmID := c.Locals("crmID").(string)
	productID := c.Params("id")

	if productID == "" {
		return c.Status(fiber.StatusBadRequest).JSON(response.CommonResponse{
			ResponseCode:    response.BAD_REQUEST,
			ResponseMessage: "Product ID is required",
		})
	}

	var body dto.UpdateProductDTO
	if err := c.BodyParser(&body); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(response.CommonResponse{
			ResponseCode:    response.BAD_REQUEST,
			ResponseMessage: "Invalid request body",
		})
	}

	product, err := pc.service.UpdateProduct(c.Context(), crmID, productID, body)
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(response.CommonResponse{
			ResponseCode:    response.BAD_REQUEST,
			ResponseMessage: err.Error(),
		})
	}

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_UPDATED,
		ResponseMessage: "Product updated successfully",
		Data:            product,
	})
}

func (pc *ProductController) DeleteProduct(c *fiber.Ctx) error {
	crmID := c.Locals("crmID").(string)
	productID := c.Params("id")

	if productID == "" {
		return c.Status(fiber.StatusBadRequest).JSON(response.CommonResponse{
			ResponseCode:    response.BAD_REQUEST,
			ResponseMessage: "Product ID is required",
		})
	}

	if err := pc.service.DeleteProduct(c.Context(), crmID, productID); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(response.CommonResponse{
			ResponseCode:    response.BAD_REQUEST,
			ResponseMessage: err.Error(),
		})
	}

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_UPDATED,
		ResponseMessage: "Product deleted successfully",
	})
}

File: frontend/app/composables/api/catalog/product.ts
import { useApi } from '../useApi'
import type { CreateProductDTO, UpdateProductDTO, CatalogItem } from '@/composables/dto/catalog/product'

interface ApiResult<T> {
  success: boolean
  data?: T
  error?: string
}

export const useProductAPI = () => {
  const api = useApi()

  const CreateProduct = async (body: CreateProductDTO): Promise<ApiResult<{ data: CatalogItem }>> => {
    try {
      const response = await api.post('/catalog/products', body)
      return { success: true, data: response.data }
    } catch (error: any) {
      return { success: false, error: error?.response?.data?.response_message || 'Failed to create product' }
    }
  }

  const ListProducts = async (params?: { search?: string; status?: string; page?: number; limit?: number }): Promise<ApiResult<{ data: { data: CatalogItem[]; total: number; page: number; limit: number; total_pages: number } }>> => {
    try {
      const response = await api.get('/catalog/products', { params })
      return { success: true, data: response.data }
    } catch (error: any) {
      return { success: false, error: error?.response?.data?.response_message || 'Failed to fetch products' }
    }
  }

  const GetProduct = async (productId: string): Promise<ApiResult<{ data: CatalogItem }>> => {
    try {
      const response = await api.get(`/catalog/products/${productId}`)
      return { success: true, data: response.data }
    } catch (error: any) {
      return { success: false, error: error?.response?.data?.response_message || 'Product not found' }
    }
  }

  const UpdateProduct = async (productId: string, body: UpdateProductDTO): Promise<ApiResult<{ data: CatalogItem }>> => {
    try {
      const response = await api.patch(`/catalog/products/${productId}`, body)
      return { success: true, data: response.data }
    } catch (error: any) {
      return { success: false, error: error?.response?.data?.response_message || 'Failed to update product' }
    }
  }

  const DeleteProduct = async (productId: string): Promise<ApiResult<null>> => {
    try {
      await api.delete(`/catalog/products/${productId}`)
      return { success: true }
    } catch (error: any) {
      return { success: false, error: error?.response?.data?.response_message || 'Failed to delete product' }
    }
  }

  return { CreateProduct, ListProducts, GetProduct, UpdateProduct, DeleteProduct }
}
